### PR TITLE
Fixed "user is not managed" error when accessing ElastiCache and MemoryDB

### LINKED ISSUE #30297: `tsh db connect` fails with ElastiCache managed user "is not manged" error
Expected behavior:
ElastiCache manged user should work

Current behavior:
```
 $ tsh db connect --db-user teleport-readonly steve-cluster-redis
localhost:55328> acl whoami
(error) ERR Teleport: Could not connect to database: failed to get password for teleport-readonly: database user teleport-readonly is not managed. Make sure that IAM auth is enabled for ElastiCache user "teleport-readonly" and Teleport database agent's IAM policy has "elasticache:Connect" permissions (note that IAM changes may take a few minutes to propagate): { "Version": "2012-10-17", "Statement": [ { "Effect": "Allow", "Action": "elasticache:Connect", "Resource": [ "arn:aws:elasticache:ca-central-1:<account>:replicationgroup:steve-cluster-redis", "arn:aws:elasticache:ca-central-1:<account>:user:*" ] } ] }
localhost:55328> 
```

Bug details:
verison: v13.3.2
introduced in https://github.com/gravitational/teleport/pull/29215 but also a bad design in managed user tracking (by database objects).

### PR DESCRIPTION
**Pull request by greedy52**
Fixes #30297 

```
$ tsh db connect --db-user teleport-readonly steve-cluster-redis
localhost:55328> acl whoami
(error) ERR Teleport: Could not connect to database: failed to get password for teleport-readonly: database user teleport-readonly is not managed. Make sure that IAM auth is enabled for ElastiCache user "teleport-readonly" and Teleport database agent's IAM policy has "elasticache:Connect" permissions (note that IAM changes may take a few minutes to propagate): { "Version": "2012-10-17", "Statement": [ { "Effect": "Allow", "Action": "elasticache:Connect", "Resource": [ "arn:aws:elasticache:ca-central-1:<account>:replicationgroup:steve-cluster-redis", "arn:aws:elasticache:ca-central-1:<account>:user:*" ] } ] }
localhost:55328> 
```

We had a few recent changes (#29215, #29864) to make copies of database objects in Database Access. This broke ElastiCache/MemoryDB managed users as the internal tracking was relying on comparing database objects by pointer instead of by name.

Changed the internal tracking to cache by `database.GetName()` and adds some extra handling when a database name is reused for a new database.

### CODE REVIEW

File: lib/srv/db/cloud/users/helpers.go [hunk 1]
```diff
@@ -80,23 +89,34 @@ func (m *lookupMap) setDatabaseUsers(database types.Database, users []User) {
...
 		}
 	}
 }
 
 // findDatabase finds the database object in provided list of databases.
-func findDatabase(databases types.Databases, database types.Database) bool {
+func findDatabase(databases types.Databases, database string) bool {
 	for i := range databases {
-		if databases[i] == database {
+		if databases[i].GetName() == database {
```
Comment (codingllama) [inline lib/srv/db/cloud/users/helpers.go:119]: Is it worth changing `databases` / `activeDatabases` to a map too? Does it get big enough to justify it?

File: lib/srv/db/cloud/users/users_test.go [hunk 2]
```diff
@@ -119,12 +119,23 @@ func TestUsers(t *testing.T) {
 		// Validate db6 is same as before.
 		requireDatabaseWithManagedUsers(t, users, db6, []string{"alice", "bob"})
 	})
+
+	t.Run("new database with same name", func(t *testing.T) {
+		newDB6 := mustCreateRDSDatabase(t, "db6")
+		users.setupDatabaseAndRotatePasswords(ctx, newDB6)
+
+		// Make sure no users are cached for "db6".
+		_, err := users.GetPassword(context.TODO(), db6, "alice")
```
Comment (codingllama) [inline lib/srv/db/cloud/users/users_test.go:128]:
```suggestion
		_, err := users.GetPassword(context.Background(), db6, "alice")
```

There's nothing "to do" here.

Same for similar occurrences.

File: lib/srv/db/cloud/users/users_test.go [hunk 3]
```diff
@@ -119,12 +119,23 @@ func TestUsers(t *testing.T) {
...
+	})
 }
 
 func requireDatabaseWithManagedUsers(t *testing.T, users *Users, db types.Database, managedUsers []string) {
 	require.Equal(t, managedUsers, db.GetManagedUsers())
 	for _, username := range managedUsers {
-		password, err := users.GetPassword(context.TODO(), db, username)
+		// Usually a copy of the proxied database is passed to the engine
+		// instead of the same object.
+		password, err := users.GetPassword(context.TODO(), db.Copy(), username)
```
Comment (codingllama) [inline lib/srv/db/cloud/users/users_test.go:138]:
```suggestion
		password, err := users.GetPassword(context.Background(), db.Copy(), username)
```

**Review by codingllama — APPROVED** [review]

**Review by GavinFrazar — APPROVED** [review]

File: lib/srv/db/cloud/users/helpers.go (hunk 1 above)
Comment (greedy52) [reply lib/srv/db/cloud/users/helpers.go]: ya, I heard some customers have hundreds of databases so definitely worth it. Updated. And actually less code =D. Thanks for the suggestion!

**Review by codingllama — APPROVED** [review]

**Pull request merged**